explorer/interpreter/type_checker.cpp

```diff
@@ -1327,49 +1363,123 @@ class TypeChecker::ConstraintTypeBuilder {
-            // `rewrite_b` will be compared against the next matching value,
-            // and so on.
+            found_existing = true;
             break;
           }
           return ProgramError(source_loc)
                  << "multiple different rewrites for `" << *rewrite_a.constant
                  << "`:\n"
-                 << "  " << *rewrite_a.unconverted_replacement << "\n"
-                 << "  " << *rewrite_b.unconverted_replacement;
+                 << "  " << *rewrite_b.unconverted_replacement << "\n"
+                 << "  " << *rewrite_a.unconverted_replacement;
         }
       }
+      if (!found_existing) {
+        new_rewrite_constraints.push_back(rewrite_a);
+      }
+    }
+    rewrite_constraints_ = std::move(new_rewrite_constraints);
+
+    // Rebuild a value with our updated rewrite constraints. Because this
+    // builder is in scope as a partial constraint type, if we perform a null
+    // substitution any nested associated constants will be rebuilt.
+    auto rebuild = [&](Nonnull<const Value*> value) {
+      Bindings null_bindings;
+      return type_checker.SubstituteImpl(null_bindings, value);
+    };
+
+    SmallRingBufferQueue<Nonnull<RewriteConstraint*>> rewrite_queue(
+        rewrite_constraints_.size());
+    for (auto& rewrite : rewrite_constraints_) {
+      if (type_checker.trace_stream_) {
+        **type_checker.trace_stream_ << "initial rewrite of "
+                                     << *rewrite.constant << " is "
+                                     << *rewrite.converted_replacement << "\n";
+      }
+      rewrite_queue.Push(&rewrite);
     }
 
-    // Rewrite `Self.X is Y` to `Replacement is Y` if we have a rewrite for
-    // `Self.X`.
-    // TODO: Properly apply rewrites throughout all the constraints. Check for
-    // cycles. This is just a very short-term hack.
-    for (auto& impl_constraint : impl_constraints_) {
-      bool performed_rewrite;
-      do {
-        performed_rewrite = false;
-        if (const auto* assoc =
-                dyn_cast<AssociatedConstant>(impl_constraint.type)) {
-          for (const auto& rewrite : rewrite_constraints_) {
-            if (ValueEqual(assoc, rewrite.constant, std::nullopt)) {
-              impl_constraint.type = rewrite.converted_replacement;
-              performed_rewrite = true;
-            }
-          }
+    // Apply rewrites to each other and find a fixed point.
+    int rewrite_iterations = 0;
+    while (auto rewrite_entry = rewrite_queue.Pop()) {
+      auto* rewrite = *rewrite_entry;
+
+      // Track that we are rewriting this rewrite.
+      RewriteInfo info = {.rewrite = rewrite};
+      currently_rewriting_ = &info;
+      auto clear_currently_rewriting =
+          llvm::make_scope_exit([&] { currently_rewriting_ = std::nullopt; });
+
+      // Rebuild the rewrite and see if it changed.
+      Nonnull<const Value*> rebuilt = rebuild(rewrite->converted_replacement);
```

Comment:
```suggestion
      // Rebuild the rewrite and see if it changed.
      currently_rewriting_ = &info;
      Nonnull<const Value*> rebuilt = rebuild(rewrite->converted_replacement);
      currently_rewriting_ = std::nullopt;
```

You only need currently_rewriting_ set for this one call, right, not anything else in scope?